src/kv/lazy-store.ts

```diff
@@ -0,0 +1,235 @@
+    }
+    this._release();
+    this._closed = true;
+  }
+
+  get closed(): boolean {
+    return this._closed;
+  }
+
+  async has(key: string): Promise<boolean> {
+    return (await this.get(key)) !== undefined;
+  }
+
+  async get(key: string): Promise<Value | undefined> {
+    let v = this._cache.get(key);
+    if (v === undefined) {
+      v = await (await this._getBaseRead()).get(key);
+      if (v !== undefined) {
+        this._cache.set(key, v);
+      }
+    }
+    return v;
+  }
+
+  private async _getBaseRead(): Promise<Read> {
+    if (!this._baseRead) {
+      this._baseRead = await this._base.read();
+    }
+    return this._baseRead;
+  }
+}
+
+class WriteImpl extends WriteImplBase implements Write {
+  private readonly _cache: Cache;
+
+  constructor(cache: Cache, read: Read) {
+    super(read);
+    this._cache = cache;
+  }
+
+  async commit(): Promise<void> {
+    // HOT. Do not allocate entry tuple and destructure.
+    const ps: Promise<void>[] = [];
+    this._pending.forEach((value, key) => {
+      if (value === deleteSentinel) {
+        this._cache.delete(key);
+      } else {
+        this._cache.set(key, value);
+      }
+    });
+    await Promise.all(ps);
+    this._pending.clear();
+    this.release();
+  }
+}
+
+type CacheEntry = {
+  value: Value;
+  size: number;
+};
+
+class Cache {
+  private readonly _cacheSizeLimit: number;
+  private readonly _shouldBePinned: (key: string) => boolean;
+  private readonly _getSizeOfValue: (v: Value) => number;
+  private readonly _pinned = new Map<string, Value>();
+  private readonly _cacheEntries = new Map<string, CacheEntry>();
+  private _size = 0;
+
+  constructor(
+    cacheSizeLimit: number,
+    shouldBePinned: (key: string) => boolean,
+    getSizeOfValue: (v: Value) => number,
+  ) {
+    this._cacheSizeLimit = cacheSizeLimit;
+    this._shouldBePinned = shouldBePinned;
+    this._getSizeOfValue = getSizeOfValue;
+  }
+
+  get(key: string): Value | undefined {
+    if (this._shouldBePinned(key)) {
+      return this._pinned.get(key);
+    }
+    const cacheEntry = this._cacheEntries.get(key);
+    if (cacheEntry) {
+      this._cacheEntries.delete(key);
+      this._cacheEntries.set(key, cacheEntry);
+    }
+    return cacheEntry?.value;
+  }
+
+  set(key: string, value: Value): void {
+    if (this._shouldBePinned(key)) {
+      this._pinned.set(key, value);
+      return;
+    }
+    const oldCacheEntry = this._cacheEntries.get(key);
+    if (oldCacheEntry) {
+      this._size -= oldCacheEntry.size;
+      this._cacheEntries.delete(key);
+    }
+    const valueSize = this._getSizeOfValue(value);
+    if (valueSize > this._cacheSizeLimit) {
+      // this value cannot be cached due to its size
+      // don't evict other entries to try to make room for it
+      return;
+    }
+    this._size += valueSize;
+    this._cacheEntries.set(key, {value, size: valueSize});
+
+    for (const entry of this._cacheEntries) {
+      if (this._size <= this._cacheSizeLimit) {
```

Comment:
My gut feeling is that it would be better to check this twice (once before the loop and once in the loop) so that we do not need to create the iterator in the common case but maybe it is too much of a premature optimization?

Reply:
its a simple enough optimization. done.